File: internship_backend_apr25/jury_solutions/go/c.go
package main

import (
	"fmt"
	"os"
	"slices"
	"strconv"
	"strings"
)
import "bufio"

type Colour struct {
	rightBorder int
	leftBorder  int
	state       State
}

type State int

const (
	New State = iota
	Open
	Close
)

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Buffer(make([]byte, 5000000), 5000000)

	var m, n int
	scanner.Scan()
	fmt.Sscan(scanner.Text(), &m, &n)
	coloursLine := make([]int, 0)
	scanner.Scan()
	line := scanner.Text()
	for _, colourString := range strings.Split(line, " ") {
		colourId, _ := strconv.Atoi(colourString)
		coloursLine = append(coloursLine, colourId-1)
	}

	colours := make(map[int]*Colour)
	for i := 0; i < m; i++ {
		colourId := coloursLine[i]
		if _, exists := colours[colourId]; !exists {
			colours[colourId] = &Colour{rightBorder: -1, leftBorder: 1000000000, state: New}
		}

		colour, _ := colours[colourId]
		if i < colour.leftBorder {
			colour.leftBorder = i
		}
		if i > colour.rightBorder {
			colour.rightBorder = i
		}
	}

	stack := make([]int, 0)
	coloursIds := make([]int, 0)

	for i := 0; i < m; i++ {
		colourIndex := coloursLine[i]
		switch colour, _ := colours[colourIndex]; colour.state {
		case New:
			stack = append(stack, i)
			colour.state = Open
		case Open:
			for coloursLine[stack[len(stack)-1]] != colourIndex {
				coloursIds = append(coloursIds, coloursLine[stack[len(stack)-1]])
				colour2, _ := colours[coloursLine[stack[len(stack)-1]]]
				colour2.state = Close
				stack = stack[:len(stack)-1]
			}
		case Close:
			fmt.Println(-1)
			return
		}
	}

	for len(stack) > 0 {
		coloursIds = append(coloursIds, coloursLine[stack[len(stack)-1]])
		stack = stack[:len(stack)-1]
	}
	fmt.Println(len(coloursIds))
	slices.Reverse(coloursIds)
	for _, colourId := range coloursIds {
		colour, _ := colours[colourId]
		id := strconv.Itoa(colourId + 1)
		leftBoarder := strconv.Itoa(colour.leftBorder + 1)
		rightBorder := strconv.Itoa(colour.rightBorder + 1)
		fmt.Println(id + " " + leftBoarder + " " + rightBorder)
	}
}

File: internship_backend_apr25/jury_solutions/cpp/c.cpp
#include <iostream>
#include <map>
#include <stack>
#include <vector>
#include <algorithm>

using namespace std;

const int NEW = 0;
const int OPEN = 1;
const int CLOSE = 2;

struct colour
{
	int right_border, left_border, state;

	colour (int _right_border, int _left_border, int _state)
		: right_border(_right_border), left_border(_left_border), state(_state)
	{
	}

	colour() = default;
};

map<int, colour> colours;

int main ()
{
	int m, n;
	cin >> m >> n;

	vector<int> colours_line;
	for (int i = 0; i < m; i++)
	{
		int c;
		cin >> c;
		colours_line.push_back(c - 1);
	}

	for (int i = 0; i < m; i++)
	{
		auto colour_id = colours_line[i];
		if (colours.count(colour_id) == 0)
			colours[colour_id] = colour(-1, INT32_MAX, NEW);

		if (i < colours[colour_id].left_border)
			colours[colour_id].left_border = i;
		if (i > colours[colour_id].right_border)
			colours[colour_id].right_border = i;
	}

	stack<int> stack;
	vector<int> colours_ids;
	for (int i = 0; i < m; i++)
	{
		auto colour_id = colours_line[i];
		switch (colours[colour_id].state)
		{
			case NEW:
				stack.push(i);
				colours[colour_id].state = OPEN;
				break;
			case OPEN:
			{
				while (colours_line[stack.top()] != colour_id)
				{
					colours_ids.push_back(colours_line[stack.top()]);
					colours[colours_line[stack.top()]].state = CLOSE;
					stack.pop();
				}

				break;
			}
			case CLOSE:
				cout << -1 << endl;
				return 0;
		}
	}

	while (!stack.empty())
	{
		colours_ids.push_back(colours_line[stack.top()]);
		stack.pop();
	}

	cout << colours_ids.size() << endl;
	reverse(colours_ids.begin(), colours_ids.end());
	for (auto colour_id : colours_ids)
	{
		cout << colour_id + 1 << " " << colours[colour_id].left_border + 1 << " " << colours[colour_id].right_border + 1 << endl;
	}

	return 0;
}
